### ISSUE TITLE
Downstream Single-cell RNA analysis with RaceID: How to get list of of filtered genes following initial processing

### ISSUE THREAD

**Issue by Phil-LoGerfo** [galaxyproject/training-material#2022]
How can I find the list of genes following the initial filtering and normalisation?

**Comment by mtekman** [comment]
This is stored in the RDS (RData) object in your history, but sadly relatively opaque to the user.

One solution is to launch an interactive Jupyter notebook, and to load the dataset (e.g. history item #7), and then extract genes using R

```R
library(RaceID)

## Load the history item
get(7)
## It tells you where it loads it:
## > '/tmp/7'

sc <- readRDS('/tmp/7')
raw.genes <- rownames(sc@expdata)
filt.genes <- rownames(sc@ndata)

## write out tables
write.csv(raw.genes, file="raw.genes")
write.csv(filt.genes, file="raw.genes")
```

*****

An update to the RaceID tools to make them more introspectable [is planned](https://github.com/galaxyproject/tools-iuc/pull/2888)

**Comment by Phil-LoGerfo** [comment]
The sc@ndata returned by filterdata() contains normalized values for all of the genes, not just the filtered genes.
The filtered genes appear to be stored in object@genes. This filtered genes list is passed to the getfdata(sc) function, which normalizes the raw reads of those filtered genes in a separate manner than that of filterdata()
The pipeline is a bit confusing because there are two normalized expression sets generated by separate normalizing methods:

**Comment by mtekman** [comment]
I definitely agree the api for this package is a little bit confusing, so I
opted to follow the same functions as those given in the vignette for
RaceID .
<https://github.com/dgrun/RaceID3_StemID2/blob/master/Reference_manual_RaceID3_StemID2.pdf>
In general, I opt to extract the gene names from the @ndata slot to avoid,
rather than calling @genes directly, since the former is taken after all
initial filtering has taken place, and the latter is populated more
ambiguously (imo).

Best,
Mehmet

**Comment by Phil-LoGerfo** [comment]
It does appear that @fdata values are equivalent to @ndata values, only
scaled up to ~0-500,000 from 0-1 range

**Comment by mtekman** [comment]
Yes, that's another slightly misleading use of slot names. In the ScanPy pipeline they make a good distinction between a normalised matrix (confounder free) and a scaled matrix (mean gene expression shifted to 0), treating them as separate entities. 

Here, this seems to be performed in one step.
